src/crh/poseidon/constraints.rs

```diff
@@ -1,279 +1,205 @@
 use super::sbox::constraints::SboxConstraints;
-use super::PoseidonRoundParams;
-use super::{Poseidon, PoseidonCRH};
+use super::{PoseidonParameters, Rounds, CRH};
 use crate::FixedLengthCRHGadget;
 use ark_ff::PrimeField;
+use ark_r1cs_std::fields::fp::AllocatedFp;
 use ark_r1cs_std::fields::fp::FpVar;
 use ark_r1cs_std::uint8::UInt8;
-use ark_r1cs_std::ToConstraintFieldGadget;
 use ark_r1cs_std::{alloc::AllocVar, fields::FieldVar, prelude::*};
 use ark_relations::r1cs::{Namespace, SynthesisError};
-use ark_std::vec::Vec;
-
-use ark_std::borrow::ToOwned;
 use ark_std::marker::PhantomData;
+use ark_std::vec::Vec;
 use core::borrow::Borrow;
 
-#[derive(Derivative, Clone)]
-pub struct PoseidonRoundParamsVar<F: PrimeField, P: PoseidonRoundParams<F>> {
-    params: Poseidon<F, P>,
+#[derive(Default, Clone)]
+pub struct PoseidonParametersVar<F: PrimeField> {
+    /// The round key constants
+    pub round_keys: Vec<FpVar<F>>,
+    /// The MDS matrix to apply in the mix layer.
+    pub mds_matrix: Vec<Vec<FpVar<F>>>,
 }
 
-pub struct PoseidonCRHGadget<F: PrimeField, P: PoseidonRoundParams<F>> {
+pub struct CRHGadget<F: PrimeField, P: Rounds> {
     field: PhantomData<F>,
-    params: PhantomData<PoseidonRoundParamsVar<F, P>>,
+    params: PhantomData<P>,
 }
 
-impl<F: PrimeField, P: PoseidonRoundParams<F>> PoseidonRoundParamsVar<F, P> {
-    fn permute(&self, input: Vec<FpVar<F>>) -> Result<Vec<FpVar<F>>, SynthesisError> {
+impl<F: PrimeField, P: Rounds> CRHGadget<F, P> {
+    fn permute(
+        parameters: &PoseidonParametersVar<F>,
+        input: Vec<FpVar<F>>,
+    ) -> Result<Vec<FpVar<F>>, SynthesisError> {
         let width = P::WIDTH;
-        assert_eq!(input.len(), width);
 
-        let full_rounds_beginning = P::FULL_ROUNDS_BEGINNING;
         let partial_rounds = P::PARTIAL_ROUNDS;
-        let full_rounds_end = P::FULL_ROUNDS_END;
-
-        let mut input_vars: Vec<FpVar<F>> = input;
-
+        let full_rounds = P::FULL_ROUNDS / 2;
```

Comment:
I think it'd be clearer if we used `P::FULL_ROUNDS / 2` in the loop

Reply:
Updated...